packages/image_picker/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerPlugin.java

```diff
@@ -85,11 +85,92 @@ public void onActivityDestroyed(Activity activity) {
     @Override
     public void onActivityStopped(Activity activity) {
       if (thisActivity == activity) {
-        delegate.saveStateBeforeResult();
+        activityState.getDelegate().saveStateBeforeResult();
       }
     }
   }
 
+  private class ActivityState {
+    private Application application;
+    private Activity activity;
+    private ImagePickerDelegate delegate;
+    private MethodChannel channel;
+    private LifeCycleObserver observer;
+    private ActivityPluginBinding activityBinding;
+
+    // This is null when not using v2 embedding;
+    private Lifecycle lifecycle;
+
+    ActivityState(
+        final Application application,
+        final Activity activity,
+        final BinaryMessenger messenger,
+        final MethodChannel.MethodCallHandler handler,
+        final PluginRegistry.Registrar registrar,
+        final ActivityPluginBinding activityBinding) {
+      this.application = application;
+      this.activity = activity;
+      this.activityBinding = activityBinding;
+
+      delegate = constructDelegate(activity);
+      channel = new MethodChannel(messenger, CHANNEL);
+      channel.setMethodCallHandler(handler);
+      observer = new LifeCycleObserver(activity);
+      if (registrar != null) {
+        // V1 embedding setup for activity listeners.
+        application.registerActivityLifecycleCallbacks(observer);
+        registrar.addActivityResultListener(delegate);
+        registrar.addRequestPermissionsResultListener(delegate);
+      } else {
+        // V2 embedding setup for activity listeners.
+        activityBinding.addActivityResultListener(delegate);
+        activityBinding.addRequestPermissionsResultListener(delegate);
+        lifecycle = FlutterLifecycleAdapter.getActivityLifecycle(activityBinding);
+        lifecycle.addObserver(observer);
+      }
+    }
+
+    ActivityState(final ImagePickerDelegate delegate, final Activity activity) {
```

Comment:
this needs javadocs.  when is this constructor preferred over the other one?

Reply:
done.